woocommerce/woocommerce-gutenberg-products-block
# Update relevant files due to L2->L0 policy change for minimum required versions.

For a number of reasons (which will be documented in a separate post on developer.woocommerce.com before merging this PR), we have decided to change the minimum version support policy for WooCommerce Blocks to L0 from L2. This effectively means that beginning with the next release of this plugin (after merging this PR), the latest version of WordPress (currently WP 5.8) and WooCommerce (currently 5.5) will be _required_.

There will be some impact on current users of the plugin but the main impact would mostly be they won't be able to update the feature plugin to the latest version until their site satisfies the minimum version requirements.

When included in WooCommerce core as a package, WooCommerce blocks will adhere to the Woo core version policy _for existing features_ and functionality already available in WooCommerce core via the feature plugin. However, _new blocks or functionality_ coming from the package inclusion _might_ be behind a WordPress version gate in keeping with the version followed when the new block/feature was being developed.

Note: This PR description will be updated with a link to the developer.woocommerce.com post about this policy change once it goes up (and I'll also be linking to the post within some of the files changed in the PR).

## Changes included

For the purpose of review, here are the files/impact of these changes:

- Plugin header, and readme files that both inform and (via built-in WordPress and WooCommerce logic) restrict the usage of the plugin.
- Main plugin file (which restricts enabling on an unsupported WordPress version). The main plugin file is not loaded when the code is loaded as a package in Woo Core so should not impact usage there.
- `src/Domain/Bootstrap.php` - Woo Blocks didn't have any restriction for Woo Core version, this was added to the bootstrap. It reads from the plugin header so we don't have to keep this updated (just the plugin header).
- Automated tests are updated to test against the supported L0 supported versions only. Unfortunately, we're unable to set things up for L2 in the automated tests because there's no way to ensure blocks is active for earlier WP versions (as far as I could tell there's no way to filter the Blocks plugin header values). I did try to do this because it's advantageous to ensure existing blocks released in Woo Core still support L2 of WordPress.
- PHPCS config. This updates the PHPCS lint rules to account for the latest version of WordPress supported by the L2 policy in core. Why not L0? We might work on code in a block or feature plugin that exists already in Woo Core and in that case this is a preventative measure against accidentally breaking that contract with the use of functions or code only existing in L0 within those things. We may have to iterate on this rule a bit if we find it leads to a lot of lint ignore rules in the codebase.

**Note: I also made some updates to the PHPUnit bootstrap so we could run the latest Woo version for PHPUnit.** In #4305, I added some logic so we'd have the ability to pin PHPUnit tests to specific Woo versions to get around errors we started seeing happening. In this PR, I was able to successfully debug the cause of the errors that was preventing us from using the latest Woo Core version and fixed it so we can now run the latest Woo Core version for PHPunit tests.

## To test

For testing the below scenarios you can manually set the "requires values" in `woocommerce-gutenberg-products-block.php` file to higher versions than what's installed in your local dev environment.

The only user facing impact of these changes is if a user is on a non-supported version of WordPress and/or WooCommerce when they are notified about a pending update. They _shouldn't_ be able to update in that scenario.

For users that update via FTP or some other direct method, they shouldn't be able to activate the plugin.

WooCommerce inclusion via package should not impact usage there (which will be tested as a part of the next package bump for WooCommerce 5.7).

## Re: GH workflow checks not completing.

Keep in mind that there are some required checks configured in the repo settings that no longer get executed due to the changes in this PR. After this PR is merged, we can update the required checks in the repo settings.
